core/primitives/address.hpp

```diff
@@ -3,16 +3,146 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-#ifndef CMAKE_HUNTER_SEED_CORE_ADDRESS_ADDRESS_HPP
-#define CMAKE_HUNTER_SEED_CORE_ADDRESS_ADDRESS_HPP
+#ifndef CPP_FILECOIN_CORE_ADDRESS_HPP
+#define CPP_FILECOIN_CORE_ADDRESS_HPP
+
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <type_traits>
+#include <vector>
+
+#include "boost/variant.hpp"
+#include "common/outcome.hpp"
 
 namespace fc::primitives {
+  using Uint = uint64_t;            // TODO:
+  using Byte = uint8_t;             // consider moving
+  using Bytes = std::vector<Byte>;  // commonly used types
+  using String = std::string;       // to a better place
+};                                  // namespace fc::primitives
+
+namespace fc::primitives::address {
+
+  /**
+   * @class Potential errors creating and handling Filecoin addresses
+   */
+  enum class AddressError : int {
+    UNKNOWN_PROTOCOL = 1, /**< Unknown Address protocol/type */
+    INVALID_PAYLOAD,      /**< Invalid data for a given protocol */
+    UNKNOWN_NETWORK       /**< Unknown network: neither testnet nor mainnet */
+  };
+
+  enum class Network : int { testnet = 0, mainnet };
+
+  enum class Protocol : int { ID = 0, SECP256K1, Actor, BLS };
+
+  template <typename T>
+  constexpr auto to_underlying(T v) noexcept {
+    return static_cast<std::underlying_type_t<T>>(v);
+  }
+
+  struct PublicKeySecp256k1Hash {
```

Comment:
This is merely a wrapper around std::vector<uint8_t> - a necessary workaround for the boost::variant to be able to differentiate between the alternatives of the same type (byte array). It's not intended to be public; the public interface will always take byte array as an input.

Furthermore, a need for such complication gives rise to a question whether a tagged union is the best way to implement 4 types of Addresses or a simple inheritance from an abstract base Address class would be more suitable. I think there is still room for discussion here because either of the approaches has their pros and cons